### PR TITLE
Webpack 4, Babel 7

### PR DESCRIPTION
**Pull request by ewandennis**
(Consider reviewing the commits in order ;) ) 

### Main Changes
 - Pulled updated CRA@2.1.3 ejected build scripts and config from a sample project
 - Reinstated our changes, including:
   - resolve modules under `./src`
   - generate tenant config files on build
   - generate snapshot of supported browsers on build
   - MDX build support for notifications
   - disable eslint on dev builds
   - build time env vars

### TODO
 - [x] Fix FacetFilter tests
 - [x] Inline critical CSS in production builds
 - [ ] Inline critical CSS in development mode
 - [x] ~Split out vendor chunk?~ Nope.
 - [x] ~Reinstate CSS classname reducer?~ Nope.
 - [x] Remove old webpack config and CSS id generation logic
 - [x] Verify functionality
 - [x] Outline newly available React features: 16.3 -> 16.7

### Testing
 - `npm start`
 - `npm run build`
 - Verify the bundle actually functions :)
 - Review build output in `build/`
 - Review bundle report in `build-reports/webpack-bundle-analyzer/index.html`
 - Serve and perf test build (instructions in `docs/perf-testing.md`)

### CODE REVIEW

File: scripts/start.js [hunk 1]
```diff
@@ -38,17 +41,41 @@ if (!checkRequiredFiles([paths.appHtml, paths.appIndexJs])) {
 }
 
 // Tools like Cloud9 rely on this.
-const DEFAULT_PORT = parseInt(process.env.PORT, 10) || 3100;
+const DEFAULT_PORT = parseInt(process.env.PORT, 10) || 3000;
```
Comment (bkemper) [inline scripts/start.js]: This will break Hamurai.

File: package.json [hunk 2]
```diff
@@ -96,75 +95,78 @@
...
     "file-loader": "^2.0.0",
+    "fork-ts-checker-webpack-plugin-alt": "0.4.14",
     "fs-extra": "^6.0.0",
-    "html-webpack-plugin": "^2.30.1",
+    "html-webpack-plugin": "4.0.0-alpha.2",
     "husky": "^1.0.0",
     "identity-obj-proxy": "^3.0.0",
     "incstr": "^1.2.3",
-    "jest": "^23.2.0",
+    "jest": "23.6.0",
```
Comment (jonambas) [inline package.json]: jest was just updated to 24 - https://jestjs.io/blog/2019/01/25/jest-24-refreshing-polished-typescript-friendly
worth bumping for the console.log swallowing alone

Comment (ewandennis) [reply package.json]: Agree. Separate PR though I think.

Comment (ewandennis) [reply package.json]: We also suffer from this issue: https://github.com/facebook/jest/issues/7704

**Comment by rajumsys** [comment]
- Getting this error 
> Uncaught Error: Incompatible SockJS! Main site uses: "1.1.5", the iframe: "1.3.0".
- Seems default config is being used for prod build. api  base paths are like `no-default-set`. _checking_

File: scripts/start.js (hunk 1 above)
Comment (ewandennis) [reply scripts/start.js]: CRA default. Fixed.

File: config/webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+            },
+            // Adds support for CSS Modules (https://github.com/css-modules/css-modules)
+            // using the extension .module.css
+            {
+              test: cssModuleRegex,
+              use: getStyleLoaders({
+                importLoaders: 1,
+                sourceMap: isEnvProduction && shouldUseSourceMap,
+                modules: true,
+                getLocalIdent: getCSSModuleLocalIdent,
```
Comment (jonambas) [inline config/webpack.config.js]: We can remove the file `generateScopedName` if we aren't using it anymore to generate class names

File: config/webpack.config.js [hunk 4]
```diff
@@ -476,7 +497,8 @@ module.exports = function(webpackEnv) {
           },
           isEnvProduction
             ? {
-                minify: {
+              inlineSource: /critical/,
```
[the remark this answers is not in the document]
Comment (ewandennis) [reply config/webpack.config.js]: Our old Webpack 3 mechanism went away so I switched from `StyleExtHtmlWebpackPlugin` to `HtmlWebpackInlineSourcePlugin` and a named `critical` chunk. It's not great and it doesn't translate well to dev mode.

I'm filing a separate bug to deal with this so we can get this out.

**Pull request merged**